subsys/shell/shell_utils.c

```diff
@@ -536,6 +536,24 @@ const struct device *shell_device_lookup(size_t idx,
 	return shell_device_internal(idx, prefix, NULL);
 }
 
+const struct device *shell_device_get_binding(const char *name)
+{
+	const struct device *dev = device_get_binding(name);
+
+	if (dev != NULL) {
+		return dev;
+	}
+#ifdef CONFIG_DEVICE_DT_METADATA
+	dev = device_get_by_dt_nodelabel(name);
+	if (dev != NULL) {
+		return dev;
+	}
+#endif /* CONFIG_DEVICE_DT_METADATA */
+
+	return NULL;
+}
```

Comment:
Thank you for your PR.
You might consider an alternative implementation like this:

```
const struct device *shell_device_get_binding(const char *name)
{
	const struct device *dev = device_get_binding(name);

	if (IS_ENABLED(CONFIG_DEVICE_DT_METADATA) && dev == NULL) {
		dev = device_get_by_dt_nodelabel(name);
	}

	return dev;
}
```

Reply:
This looks cleaner IMO.

Reply:
Done.
Honestly, I was sure this wouldn't work since `device_get_by_dt_nodelabel` is only declared in `include/zephyr/device.h` when `CONFIG_DEVICE_DT_METADATA` is defined. I assumed it wouldn't compile when `CONFIG_DEVICE_DT_METADATA` isn't enabled.
But then it did!  How does this work? Does it get optimized out by the compiler?

Reply:
I am guessing it is because of this:
```
config SHELL_DEVICE_HELPERS
	bool "Shell device helpers"
	imply DEVICE_DT_METADATA
	default y if !SHELL_MINIMAL
```
I anticipate there might be some warnings when the `SHELL_MINIMAL` is set and the `__DOXYGEN__` is not defined.

Reply:
No I added `CONFIG_DEVICE_DT_METADATA=n` to `prj.conf` and verified that it changed in `build/zephyr/.config` and it still compiled with no warnings

Reply:
Does it still compile if you disable compiler optimization?

Reply:
If you mean `CONFIG_NO_OPTIMIZATION` then yes. Just checked.